Memory initializer in string literal

gagern opened the pull request:
Having the memory initializer as a separate file can become a huge problem in some environments. It would be nice if we could offer more alternatives. I see how big array literals would be kind of bulky. But how about encoding the memory initializer into one or more strings? Here is some code to do that, based on a base 88 encoding scheme. If you like the idea, it would be nice if you could introduce some setting to switch between file and string.

gagern commented:
I just found out about #2188 and the comparisons done there, and I think there are better approaches than the one I used here.

gagern commented:
Since base 88 is not as memory-efficient as hoped, and in particular compresses really poorly, here is a forced push using a lookup table instead. This follows findings reported on #2188. I've by now included a setting as well, and only emit mem file *or* initializer string, not both. So in my opinion this is now ready for review and merge.

kripken commented:
The first issues here are as discussed in that other pull - we need to make sure this works with all other features. In particular we had problems there with minification breaking the string, I don't remember the details though. If the full test suite passes, then we are probably ok.

Aside from that, I saw some numbers in the other pull, but can you please summarize how this works, the measured effect on code size, and the measured effect on startup time?

gagern commented:
Do you mean the string minification in this code, or the one applied by e.g. closure afterwards? The former was discussed and addressed in https://github.com/kripken/emscripten/issues/2188#issuecomment-41489929 if that is what you have in mind.

Using the tool from https://github.com/gagern/Web-Benchmarks/tree/tinylut/meminit I get the following figures for memory efficiency, sorted by gzipped size, and augmented by some load time experiments:

| Representation | uncompressed size | gzipped size | Firefox | Chrome | Safari |
| -------------- | ----------------- | ------------ | ---- | ---- | ---- |
|         binary |          167.2 KB |      42.2 KB |
|     **tinylut** <== |          187.4 KB |      46.2 KB | 5.4 ms | 7.7 ms | 73.1 ms |
|        minstr ^ 32 |          217.9 KB |      46.9 KB |
|         minstr |          230.8 KB |      47.1 KB | 4.6 ms | 8.1 ms | 89.6 ms |
|            str |          312.1 KB |      48.6 KB |
|           int8 |          536.5 KB |      55.7 KB | 30.4 ms | 64.7 ms | 54.4 ms |
|         base64 |          223.0 KB |      59.6 KB |
|         base88 |          186.7 KB |      61.3 KB |
|          int32 |          364.3 KB |      79.7 KB |

The current mem init file based approach depends not only on parse time but on network latency, so it's hard to include that in the table. The load times above were the median of 16 runs each.

I'm currently running the test suite on the plain incoming branch, starting not with the sanity component but the default component. Once that succeeds I'll run tests for my pull requests.

gagern commented:
As for how this works: I compute how often each byte occurs, and then assign printable values to the common bytes, leaving the values which require escape sequences for the less common values. Small characters are encoded in octal notation if possible, since that's shorter. To reconstruct, I provide a translation table with 256 elements. Each char code from the string is looked up in that table to translate it into a byte which is then stored in some typed array.

kripken commented:
What does "lut" mean in "tinylut"?

kripken commented on emcc:
```diff
@@ -708,7 +708,7 @@ try:
   if llvm_opts is None: llvm_opts = LLVM_OPT_LEVEL[opt_level]
   if opt_level == 0: debug_level = max(3, debug_level)
-  if memory_init_file is None: memory_init_file = opt_level >= 2
+  if memory_init_file is None: memory_init_file = 2 if opt_level >= 2 else 0
```
wouldn't this change make this new mode the default in `-O2+`? But it's less optimized than a binary mem init?

gagern replied:
Yes, it would make this the default. I would have assumed that for most applications, the overhead for a second request would be bigger than the performance cost of parsing that initializer. But I guess you know your customers better than I do, so if you say that for most their setups mem files are preferable, then I won't argue with that. I know that I personally will try to use the string initializer simply because I prefer to have a single artifact containing the whole program, but that's just a matter of taste to me.

kripken replied:
It is possible to do the requests in parallel. But I didn't really consider until now that if this is efficient enough, it could be better than even that. I wonder how to measure it though.

Regardless, until we have such measurements, I don't think we should change the default - we shouldn't confuse users with changes unless they have proven benefits.

gagern commented:
“LUT” = LookUpTable, the one which translates characters back to byte values.

gagern commented on emcc:
```diff
@@ -1385,7 +1386,7 @@ try:
 
   if memory_init_file:
     if shared.Settings.USE_TYPED_ARRAYS != 2:
-      if type(memory_init_file) == int: logging.warning('memory init file requires typed arrays mode 2')
```
Dropping this typ echeck means that I'll now have greater chances of hitting this message. I guess I should introduce some other variable to distinguish a user-requested setting from an automatically chosen one. Or should I use extra values for that single variable?

kripken replied:
we have actually deprecated non typed arrays mode 2 code, so that should never be hit

chadaustin commented:
Hi @gagern.  I am very happy that you are looking into this.  We have wanted similar behavior for a long time.

However, I would very much prefer that the string be an as-is JavaScript string, such that the decoding loop is literally:

for (var i = 0; i < memInitStr.length; ++i) {
    HEAP8[base+i] = memInitStr[i];
}

Large regions of contiguous zeroes could be elided.

This way the data will be trivially inspectable with human eyes, and my understanding (though prove me wrong) is, the code size hit after gzip is rather small.

gagern commented:
@kripken: The default for `-O2` is now back to `--memory-init-file 1` so this thing can be evaluated first, and may become the default later on if it proves useful. One drawback is that now the test suite won't test this at all, right? Can you tell me how much test coverage you'd want for this? Is one case working with the new code enough, or do you want several tests compiled with `--memory-init-file 2` as well? Can you give me some pointers about where to start adding such tests?

@chadaustin: The string is now literal, with no table lookup step, so you can read the memory content. (By the way, I notice that libc++ content contains a lot of full paths to source files, thus exposing the developer's directory structure. I wonder what to do about that.)

So far my things are untested, because I have trouble getting a recent fastcomp installed. Apart from kripken/emscripten-fastcomp#85 it's trying to link against libedit, which I don't have installed. Working on this, may file a separate ticket if I can't resolve it.

I had to resolve quite a number of conflicts to keep up with recent development. Should I rebase my changes?

waywardmonkeys commented:
Between @kripken and I, it looks like you've probably had a lot of changes to keep up with.

It looks like this change would be a lot simpler if you:

* Stored the data using the same variable name (``memoryInitializer`` instead of ``memInitString``)
* If you want to have an option to flip between encodings, just add a new setting:  ``MEMORY_INITIALIZER_ENCODING`` which can default to 1 (array) but have an option for 2 (string) or other things in the future. This avoids the confusing overloading of ``mem_init_file`` which ends up bloating the changes a good bit.

kripken commented:
+1 to @waywardmonkeys suggestions

Also: yes, need to rebase. For tests, we should probably use this on a whole test mode, like asm2f (see end of `test_core.py`).

Now that this is a string without a table lookup, where is it in the numbers listed before (which line is it there)?

kripken commented:
See memory init testing in `runner.py` (grep for `memory-init-file`), we could add something similar for when the string is used.

Btw, we have a few settings that disable the memory init file (like shared libraries). We might want to default them to this new mode, perhaps.

gagern commented:
I rebased my changes, and hope this makes sense the way I did things. Running the tests will take some time yet, since I need to update all my toolchain first, where I'm encountering a bunch of problems, and it's pretty late in the day here.

I noticed that at some later point, we probably should attempt to exempt the string literal from the influence of the closure compiler, since it will re-encode the string to something which is longer again. The semantics stay the same, though, so this is no crucial change. And I haven't yet figured out how exactly the asm.js output gets protected from closure, to see whether we could duplicate that.

gagern commented:
I've got some failing tests in a core asm2m run: test_python test_the_bullet test_files test_fnmatch test_strings test_webidl error or fail. I'm investigating the first of these.

Something appears to be post-processesing the string literal, turning byte escapes into unicode escapes but also turning octal escapes into their binary equivalent. I haven't figured out yet who's doing this. But it caused me to read [section 7.8.4 of the ECMA 262 spec](http://www.ecma-international.org/ecma-262/5.1/#sec-7.8.4) and find out that indeed almost any unicode character may be contained in a string literal, while on the other hand support for octal escapes is optional and even forbidden in strict mode. Does that mean that we should avoid escaping anything except for newlines and quotes? Even a two-byte UTF-8 sequence for +U0080 through +U00FF is shorter than the four-byte `\x??` sequence. Can we rely on the input being UTF-8?

In any case, the problem why the test is that the hex-to-octal rewrite would also affect things like `\\x01` i.e. where the backslash itself is part of the string content. This completely breaks the memory alignment. That will need fixing. But if we have to drop all support for octal escapes in any case, then that replacement will no longer be a problem either.

I think I'll push a change where I rely on the source being parsed as UTF-8. Let me know whether you think such an assumption acceptable. With that change applied, some of the tests that used to fail now pass, but some still fail. `test_the_bullet` complains about `no input files`; I've seen this before in #3328. `test_files` assumes that we should have a mem file, will need modification. Likewise `test_fnmatch` and `test_strings` assume something about the presence of a memory initializer, will need modification. `test_webidl` says `No such file or directory: 'WebIDLGrammar.pkl'`, not sure yet why.

waywardmonkeys commented:
While you were away, I had tried implementing this as well. My implementation ended up going into a different part of the code, but it kept running into problems as well. There's something about this that is hard to crack!

gagern commented:
@waywardmonkeys: Do you have your approach in some branch here on github? What kinds of problems did you encounter?

gagern commented:
OK, I think I fixed the previously failing test cases. I dropped some test functionality in c33338387d217f17dfb6cc6da68305e074dfc19b, please have a look whether that's acceptable. I also had to merge 1.33.1 to avoid conflicts with current incoming due to #3266 (very interesting development there, by the way). Now I'm updating my emsdk to current incoming, in order to run tests once more. Again.

Once thing I've been thinking about, regarding the fact that the string literal now depends on the character encoding used for the source file: perhaps we should include some kind of checksum with the string literal, in order to at least detect any corruption due to wrong encoding there. Should I add some CRC32 or Adler32 or similar? Should the verification be made optional inside an `#if ASSERTIONS` block?

kripken commented:
A checksum sounds like a very good idea, and yes, in an `ASSERTIONS` block.

When you say UTF8 above, is the requirement that the entire JS file - including `--pre.js`, js libraryes etc. that the user provided - must be UTF8, and not another encoding? If so, I'm not sure offhand how serious a problem it is, but it doesn't sound like something obviously ok to do.

gagern commented:
What I mean is that the browser must decode it as UTF-8, probably because the HTTP server delivered it as such. If that doesn't sound good, we should go with `\x??` escapes. If things get minified, the result will for some reason have unicode escapes in any case. I just manged to understand CRC32 well enough that I don't have to ship a table but can generate that on the fly. Working on this.

gagern commented:
`tests/runner.py asm2m` only complains about `test_the_bullet` which it always does on my system. Looks good to me. What do you think?

gagern commented:
I finally found [the uglify call](https://github.com/kripken/emscripten/blob/3b49f69930b1ecea44002c934165ac11c7639b44/tools/js-optimizer.js#L167-174) which explicitely passes `ascii_only: true` as an option to uglify-js. So the hex escapes I'm now using above are pretty much in line with that. if at some point we choose to handle either of these situations differently, allowing binary data in UTF-8 encoding or whatever, then it would make sense to handle both situations (with and without uglify) the same.

Since mishoo/UglifyJS2@81f5efe39a6d44d3037dda8bcff01195f5699139 uglify knows how to produce `\x??` escapes instead of `\u????`, so we might want to update our dependency there, to keep post-uglify string literals reasonably short.

kripken commented on emcc:
```diff
@@ -1340,14 +1344,42 @@ try:
       if len(s) == 0 and not shared.Settings.EMTERPRETIFY: return m.group(0) # emterpreter must have a mem init file; otherwise, don't emit 0-size ones
-      open(memfile, 'wb').write(''.join(map(lambda x: chr(int(x or '0')), s.split(','))))
+      membytes = [int(x or '0') for x in s.split(',')]
+      if not shared.Settings.EMTERPRETIFY:
```
was this needed because the emterpreter used to append its data to the normal memory init file? that is no longer true, so this should be removable.

gagern replied:
This was mostly because if the memory is all null bytes, then I do a `return ''`, which I considered semantically equivalent to `return m.group(0)`. Come to think of it, if it indeed is, then perhaps I should make the two lines agree. Can I use `return ''` in both cases? Anyway, the `return m.group(0)` is *not* used for `EMTERPRETIFY`, presumably because we need a mem file there, even if it's empty. If it is OK for the emterpreter to have trailing zeros stripped, then I can move that condition to the `if not membytes` case.

kripken replied:
actually, all the emterpreter stuff here (and around it) is all no longer needed (there is no interaction between the emterpreter and mem init file anymore). removing any special-casing for it should be fine, if that makes your changes simpler.

kripken commented on emcc:
```diff
@@ -1340,14 +1344,42 @@ try:
+          membytes.pop()
+        if not membytes:
+          return '';
+      if not memory_init_file:
```
this section needs a comment

kripken commented on emcc:
```diff
@@ -1340,14 +1344,42 @@ try:
+        if not membytes:
+          return '';
+      if not memory_init_file:
+        crcTable = []
```
we only need the crc if `shared.Settings.ASSERTIONS`

kripken commented on emcc:
```diff
@@ -1340,14 +1344,42 @@ try:
+        s = s.replace('\n', '\\n').replace('\r', '\\r')
+        def escape(x): return '\\x{:02x}'.format(ord(x.group()))
+        s = re.sub('[\x80-\xff]', escape, s)
+        return "var memoryInitializer = '%s';" % s
```
we use chunking on the array, because (1) view source in browsers can fail on a massively long line, and (2) js engines would run out of stack trying to parse it.

we should investigate if a massively long string can be parsed by js engines (just by writing a little program that generates such a js program, without emscripten)

gagern replied:
It took me a while to figure out that the call to `optimize_initializer` does that splitting, but only if we didn't use a memory initializer file or string. I could implement a similar splitting for the strings, in which case I'd probably turn the variable assignment into a function call, similar to the `alloc` call it's replacing. But perhaps we should only do so if it's really needed.

Investigating that need, do you have a one-off test in mind, or a part of the browser test suite? What browsers are officially supported? What kind of memory initializer size do you have in mind? Out of the blue, I'd try 32M and assume that if this works, any reasonable program will work.

Viewing the source could be hell anyway, as it currently stands, due to the non-printable control bytes included in the string literals. If that's a problem, we'll have to use way more hex escape sequences, and perhaps try to convince uglify to do something alike. I had the impression that small file sizes were more important here.

kripken replied:
a one-off test is what i was thinking of. just to see if there are surprising issues on limits on string literal sizes.

32M sounds fine to test. I would try recent firefox and chrome. others too if its convenient.

kripken commented on emcc:
```diff
@@ -1289,6 +1289,10 @@ try:
   extra_args = [] if not js_libraries else ['--libraries', ','.join(map(os.path.abspath, js_libraries))]
+  if memory_init_file:
+    shared.Settings.MEM_INIT_METHOD = 1
+  elif shared.Settings.MEM_INIT_METHOD == 1:
```
Instead of silently doing this, let's assert that `MEM_INIT_METHOD` should not be 1 (if not `memory init file`)

kripken commented on src/postamble.js:
```diff
@@ -1,6 +1,32 @@
+    crc = table[(crc ^ s.charCodeAt(i)) & 0xff] ^ (crc >>> 8);
+  assert(crc === 0, "memory initializer checksum");
+#endif
+  for (i = 0; i < n; ++i)
```
curly braces

kripken commented on src/postamble.js:
```diff
@@ -1,6 +1,32 @@
+  var crc, bit, table = new Int32Array(256);
+  for (i = 0; i < 256; ++i) {
+    for (crc = i, bit = 0; bit < 8; ++bit)
+      crc = (crc >>> 1) ^ ((crc & 1) * 0xedb88320);
```
curly braces

kripken commented on tests/test_core.py:
```diff
@@ -7387,6 +7387,7 @@ def setUp(self):
 asm2g = make_run("asm2g", compiler=CLANG, emcc_args=["-O2", "-g", "-s", "ASSERTIONS=1", "-s", "SAFE_HEAP=1"])
 asm1i = make_run("asm1i", compiler=CLANG, emcc_args=["-O1", '-s', 'EMTERPRETIFY=1'])
 asm3i = make_run("asm3i", compiler=CLANG, emcc_args=["-O3", '-s', 'EMTERPRETIFY=1'])
+asm2m = make_run("asm2m", compiler=CLANG, emcc_args=["-O2", "--memory-init-file", "0", "-s", "MEM_INIT_METHOD=2"])
```
this is also a list of these modes in `tests/runner.py` that this new one needs to be added to (search for e.g. `asm3i`)

kripken replied:
also `tests/parallel_test_runner.py` (can put it after `asm2`)

kripken replied:
we need to test the combination of this option and ASSERTIONS, to check that the assertions do not fire. Can probably keep it on for the entire test mode, but should also test generating code without it (see e.g. `test_cubescript` which does that for a different mode).

it would also be good to get a test showing the assertion firing when the crc is wrong. in `tests/test_other.py`, can emit a file, modify the string manually, then run it and see the error.

kripken commented:
I have a bunch of notes, but overall this looks quite good.

After we get this in, we can do more measurements and consider having it on by default.

kripken commented:
Testing-wise, I think we should add a test that uses the full range of values in the memory initializer. Say all pairs of bytes (so 2^16 pairs). The testcase could be generated at runtime (creating structs in the global scope in C, something like that).

gagern commented on tests/runner.py:
```diff
@@ -252,16 +260,10 @@ def build(self, src, dirname, filename, output_processor=None, main_file=None, a
         # side memory init file, or an empty one in the js
         assert ('/* memory initializer */' not in src) or ('/* memory initializer */ allocate([]' in src)
-      else:
-        assert 'memory initializer */' in src or '/*' not in src # memory initializer comment, or cleaned-up source with no comments
```
@kripken: Judging from your lack of comment on this line, I take it that dropping this particular check is OK?

kripken replied:
yes.

gagern commented:
Testing all possible pairs makes a lot of sense. A generated sequence of string literals should do the trick there, far shorter than global structs. It might make sense to also make this part of the browser test suite, to ensure that all browsers can handle the emitted code. And to use closure-compiled and uglified versions in addition to a plain one.

kripken commented:
Yes, that testing sounds good.

gagern commented:
I incorporated your suggestions. I've included a browser test to ensure that large strings are no problem. So far (recent Firefox and Chromium on Linux) they are not, so I'd not implement chunking just now. The closure compiler takes pretty long for that test case, so if you are concerned about execution time for the test suite, you might want to drop the closure run of `test_mem_big`.

I've got some browser test cases fail for me, but apparently for unrelated reasons: 

* ERROR: test_emscripten_api
* ERROR: test_glgears_proxy
* ERROR: test_s3tc_crunch
* ERROR: test_s3tc_crunch_split
* ERROR: test_sdl2_image_compressed
* ERROR: test_sdl_image_compressed
* FAIL: test_compressed_file
* FAIL: test_compression
* FAIL: test_emrun

It might make sense to double-check these before merging this request, just to be sure I didn't misinterpret the situation. If you want me to, I'll file a separate ticket to investigate these issues in detail even if they are unrelated to the issue at hand.

kripken commented:
No need to file, that sounds good.

kripken commented on emcc:
```diff
@@ -1340,18 +1344,48 @@ try:
   js_transform_tempfiles = [final]
 
-  if memory_init_file:
+  if memory_init_file or shared.Settings.MEM_INIT_METHOD == 2:
```
this check could be `MEM_INIT_METHOD > 0`? seems cleaner

kripken replied:
Up to you though, I don't feel strongly about this one.

kripken commented on emcc:
```diff
@@ -1340,18 +1344,48 @@ try:
+      membytes = [int(x or '0') for x in s.split(',')]
+      while membytes and membytes[-1] == 0:
+        membytes.pop()
+      if not membytes and not shared.Settings.EMTERPRETIFY:
```
i would like to not see new hacks put in for the emterpreter. i think this is not necessary here, in fact. is it?

kripken commented on emcc:
```diff
@@ -1340,18 +1344,48 @@ try:
+      s = m.group(1)
+      if len(s) == 0 and not shared.Settings.EMTERPRETIFY: return '' # emterpreter must have a mem init file; otherwise, don't emit 0-size ones
+      membytes = [int(x or '0') for x in s.split(',')]
+      while membytes and membytes[-1] == 0:
```
this is a new optimization?

gagern replied:
Yes, stripping trailing zeros is new. At least in this context; don't know whether the `split_initializer` code affects trailing zeros as well.

kripken commented:
Looks great aside from a few final questions.

gagern commented:
Incorporated, hope this is ready now.

kripken commented:
Looks good. I'm doing some local testing before merging. I am seeing the string literal be about as large, code-size, as a mem init file (when adding the mem init file to the js), and no mem init file is 10% smaller than both. Is that expected? Seems a little surprising to me. (This is on BananaBread.)

I also see the string literal as fastest for startup, mem init after, and no mem init file as slowest, which makes sense.

kripken commented:
gzipped sizes are all quite close. mem init is smallest, them string literal, then nothing.

gagern commented:
> I am seeing the string literal be about as large, code-size, as a mem init file (when adding the mem init file to the js)

Expected: a byte which is smaller than 0x80 usually encodes using a single byte, the others require four bytes, but there are fewer of those in most applications. Plus the initialization code is much smaller for the literal.

> no mem init file is 10% smaller than both. Is that expected? Seems a little surprising to me.

This sounds really strange!!! I dropped the mem init handling code for those cases, which might account for some size reduction, but since a byte of memory gets encoded by two to four bytes for the array literal, that change should soon become negligible.

> I also see the string literal as fastest for startup, mem init after, and no mem init file as slowest, which makes sense.

Depends on connection and browser. For slow connections and no cached file, the time between when the page is loaded and when the code can start may be much larger with mem init files. Of course, otherwise the JavaScript will take long to load, so the effect will likely be reduced to how long it takes from start of the script execution till when the mem data starts coming in.

> gzipped sizes are all quite close. mem init is smallest, them string literal, then nothing.

Very much expected.

Does this mean you're about ready to merge, when your tests are concluded?

kripken commented:
Ok, one reason the nothing (no mem init, no string init) method is smaller, is that we split the init into chunks. And we can then wipe out runs of 0 at the end of each chunk. This also explains why after gzip, that benefit vanishes.

It might make sense to do that with the string init as well. Perhaps a refactoring could reuse that code in `optimize_initializer` in `tools/shared.py`.

Another possible followup could be to move the new `emcc` code (crc, etc.) here into a method in `tools/shared.py`, alongside `optimize_initializer`.

Anyhow, two thoughts for possible followups. I am ready to merge this now, but I see the commits are not squashed, and there is a merge commit which we should not have?

gagern commented:
The zero runs explain this change. But this kind of optimization may well make the file bigger after gzip, so the tradeoff should be evaluated carefully.

Do you want all commits squashed into one? Or is it enough if I rebase to 1.33.1 to avoid the merge? That was neccessary because of conflicts in the preamble between my changes and the pthreads code. The question mark at the end of your post make it slightly unclear whether you are making a request or asking my opinion, and personally I like to see the evolution of a feature in the history.

kripken commented:
Yeah, I'm ok with history, as long as it is informative. Merge commits should be rebased out, but if you feel the rest of the history has value, that's fine.

gagern commented:
I've squashed some commits, but kept others apart where I considered the extra history valuable. And I verified that my final result is equivalent to the 4db4ee6b68911e09bfdca80680143f99be167143 you reviewed. I'd say everything should be ready for merge now.

chadaustin commented:
I am so so so happy this is happening.  Thanks for all the hard work.

kripken commented:
Merged. Thanks!

I also pushed a commit to refactor the crc code into `tools/shared.py`.

gagern commented:
Thanks for the cooperation, I'm looking forward to using emscripten with this new improvement. Do you have any idea how to gather feedback on whether this should become the default for some settings?

kripken commented:
Opening a thread on the mailing list might be a good start, asking people to test it out.

Meanwhile I ran some fuzzing on this last night, and found no issues.

kripken commented:
It looks like there is windows-only breakage on the bots due to this. For example `asm2.test_align_moar` emits this: https://dl.dropboxusercontent.com/u/40949268/emcc/bugs/tmp_test_align_moar.js . See the end, it looks like the mem init is not escaped properly perhaps.

juj commented:
On OS X, I see that when looking at the output of `asm2m.test_align_moar` in the `emscripten_temp/` directory (run with `EM_SAVE_DIR=1 EM_BUILD_VERBOSE=3 python tests/runner.py asm2m.test_align_moar`), that the generated .js file contains some null bytes in the middle of the file. I think we shouldn't put any null bytes in the file, and also I wonder how bytes 1-31 work.

Although I think the issue of the test crashing optimizer.exe on Windows is not in the null bytes, but that somehow python is computing the file write bad, and it's outputting/truncating the rest of the file somehow, since nothing gets printed to the file after the memory initializer is outputted.

Was the design of the string memory initializer to depend on outputting bytes in the range 0-31 raw to the generated .js file? Or should those have gotten escaped somehow?

gagern commented:
> Was the design of the string memory initializer to depend on outputting bytes in the range 0-31 raw to the generated .js file? Or should those have gotten escaped somehow?

Originally I had intended to escape control bytes, then I noticed that our version of uglify [would unescape those](https://github.com/kripken/emscripten/blob/master/tools/eliminator/node_modules/uglify-js/lib/process.js#L1362-1373), [read in the spec](https://github.com/kripken/emscripten/pull/3326#issuecomment-106743344) that almost any code point is valid in a string literal, including nulls, and decided to go with that myself.

Should I write a pull request to always escape control bytes, and we investigate relaxing that later on? Or should we investigate now until we understand the problem, then fix just that and hope all other platforms are fine?

> Although I think the issue of the test crashing optimizer.exe on Windows is not in the null bytes, but that somehow python is computing the file write bad, and it's outputting/truncating the rest of the file somehow, since nothing gets printed to the file after the memory initializer is outputted.

Perhaps Python should open the file in binary mode, i.e. `'wb'` instead of `'w'` as an argument? I thought that would only affect line terminator translation, but I seem to recall that windows has an EOF control byte, and perhaps that's what's interfering here. It might be worthwhile to just try writing all possible bytes in python, see whether that triggers the problem.

```py
open('foo.bin', 'w').write(''.join(map(chr,range(256))))
```

Unfortunately I don't have a Windows at hand just now, so can someone on Windows please give the above a try and report back? From reading [Wikipedia](https://en.wikipedia.org/wiki/End-of-file#EOF_character) I'd guess 26 or 4 to be the most likely bytes to cause problems here.

juj commented:
I'm now seeing one crossplatform difference, that causes the breakage. If one runs the following python program

	f = open('foo.txt', 'w')
	f.write('hello text file!!!\n')
	f.write(''.join(map(chr,range(256))))
	f.write('good bye text file!!!\n')
	f.close()

	contents = open('foo.txt', 'r').read()
	print contents

on Windows, the second read will truncated on what looks like the byte 0x19 (end of medium) or 0x20 (substitute), and the application prints

	hello text file!!!
	 ☺☻♥♦
	♫☼►◄↕‼¶§▬↨↑↓

but on OS X, the read will complete past each byte, and the application prints

	hello text file!!!
	 ☺☻♥♦
	♫☼►◄↕‼¶§▬↨↑↓→←∟↔▲▼ !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]
	^_`abcdefghijklmnopqrstuvwxyz{|}~⌂ÇüéâäàåçêëèïîìÄÅÉæÆôöòûùÿÖÜ¢£¥₧ƒáíóúñÑªº¿⌐¬½¼¡
	«»░▒▓│┤╡╢╖╕╣║╗╝╜╛┐└┴┬├─┼╞╟╚╔╩╦╠═╬╧╨╤╥╙╘╒╓╫╪┘┌█▄▌▐▀αßΓπΣσµτΦΘΩδ∞φε∩≡±≥≤⌠⌡÷≈°∙·√ⁿ²
	■ good bye text file!!!

The effect is that writing the memory string initializer succeeds correctly, but afterwards reading it back with `open(jsfile).read()` or `open(jsfile, 'r').read()` truncates the read data at the first seen null byte. On Windows, `open(jsfile, 'rb').read()` does properly read all the bytes.

However even if I go in places and replace 'r' with 'rb', I am still getting corrupted output in the final executable. I am not sure if I found all the relevant places to transform, or if I missed a place. The output does change from optimizer.exe crashing, to another kind of corruption where the end of the memory initializer reads `\u00c0=(function(global,env,buffer) {`, i.e. it is missing the terminating " character for the string, and the subsequent code is joined directly into it, generating an invalid .js file. Unfortunately I'm running out of time here to debug this, but I believe the root cause of these issues is with crossplatform inconsistencies in ascii text file handling.

kripken commented:
I don't think there is a rush to fix this, in that the breakage is only on the string initializer mode, which is off by default. The only thing this blocks is us moving to that mode by default. But since that would be good to do, this is important to fix, I'm just saying I think we can wait for a proper fix, no need for an interim workaround.

It does sound like binary/text issues are in play here. Possibly 'rb'/'wb' in the right places would fix it, as you two suggest. I don't have a windows machine either, but perhaps someone else reading this does. We could also ask on irc or the mailing list. I'd also be ok to merge a pull that has those fixes, and we'll see what happens on the windows bot.

Side note, I've been wondering why the new `test_align_moar` triggered this issue, when it wasn't seen earlier. I realized that there is something special about it - the memory init file begins with a bunch of zeros, around 8, because it has to add them as padding to fix alignment. It's possible not many other tests have such a run of zeros, although that is surprising.

gagern commented:
> the application prints
>
>     hello text file!!!
>      ☺☻♥♦
>     ♫☼►◄↕‼¶§▬↨↑↓

OK, what we are seeing here is apparently some raw bytes decoded according to [code page 437](https://en.wikipedia.org/wiki/Code_page_437) or something very similar. We have bytes 00 through 04 in the first row. Not sure where the rest of that line went, but the next line starts with 0E, so I'd say 0D got interpreted as a carriage return without line feed, so stuff between the beginning of the file and that position got overwritten. Then we have bytes 0E through 19 in the second row, but no 1A. This is consistent with 1A denoting end of file in text mode. So my hunch there was correct.

> The output does change from optimizer.exe crashing, to another kind of corruption where the end of the memory initializer reads `\u00c0�=(function(global,env,buffer) {`, i.e. it is missing the terminating " character for the string, and the subsequent code is joined directly into it, generating an invalid .js file.

I have no idea yet where that may come from.

> I don't have a windows machine either, but perhaps someone else reading this does.

I do have a VM with Windows. It's painfully slow, and I have nearly no useful tools installed there, but if everything else fails, I guess I can set up an emsdk there.

> Side note, I've been wondering why the new test_align_moar triggered this issue, when it wasn't seen earlier.

That is strange indeed. As I said, I guess it's not the null bytes, but the 1A bytes instead. But `test_meminit_pairs` at the very least should have had that as well. Could it be that that test case does not cover as much as we expect it to cover? One more reason to investigate the issue properly, instead of just working around the symptoms.

kripken commented:
I'm going to disable the `asm2m` tests for now, until we resolve this. It breaks the windows bot, and might be hiding other breakage as development continues on incoming.

gagern commented:
If you want to, you can try the following change instead:

```diff
-s = re.sub('[\x80-\xff]', escape, s)
+s = re.sub('\x00(?![0-9])', '\\0', s)
+s = re.sub('[\x00-\x1f\x7f-\xff]', escape, s)
```

Can't write a full pull request just now. The idea is to turn any null byte which doesn't precede a digit into `\0` and escape control characters along with the non-7bit-ASCII bytes.

kripken commented:
Thanks, I don't have time either currently, though (no windows machine, and busy with the syscalls landing).

aidanhs commented:
> However even if I go in places and replace 'r' with 'rb', I am still getting corrupted output in the final executable. I am not sure if I found all the relevant places to transform, or if I missed a place. The output does change from optimizer.exe crashing, to another kind of corruption where the end of the memory initializer reads \u00c0�=(function(global,env,buffer) {, i.e. it is missing the terminating " character for the string, and the subsequent code is joined directly into it, generating an invalid .js file. Unfortunately I'm running out of time here to debug this, but I believe the root cause of these issues is with crossplatform inconsistencies in ascii text file handling.

Emscripten really should use 'rb' everywhere anyway.

waywardmonkeys commented:
> Emscripten really should use 'rb' everywhere anyway.

:+1:

kripken commented:
If we can confirm that fixes this issue, and has no downsides (I don't know windows enough to tell...), sounds good to me.
